code/controllers/subsystem/spatial_gridmap.dm

```diff
@@ -0,0 +1,567 @@
+#define GRID_CELL_ADD(cell_contents_list, movable_or_list) \
+	if(!length(cell_contents_list)) { \
+		cell_contents_list = list(); \
+		cell_contents_list += movable_or_list; \
+	} else { \
+		cell_contents_list += movable_or_list; \
+	};
+
+#define GRID_CELL_SET(cell_contents_list, movable_or_list) \
+	if(!length(cell_contents_list)) { \
+		cell_contents_list = list(); \
+		cell_contents_list += movable_or_list; \
+	} else { \
+		cell_contents_list |= movable_or_list; \
+	};
+
+//dont use these outside of SSspatial_grid's scope use the procs it has for this purpose
+#define GRID_CELL_REMOVE(cell_contents_list, movable_or_list) \
+	cell_contents_list -= movable_or_list; \
+	if(!length(cell_contents_list)) {\
+		cell_contents_list = dummy_list; \
+	};
+
+/**
+ * # Spatial Grid Cell
+ *
+ * used by [/datum/controller/subsystem/spatial_grid] to cover every z level so that the coordinates of every turf in the world corresponds to one of these in
+ * the subsystems list of grid cells by z level. each one of these contains content lists holding all atoms meeting a certain criteria that is in our borders.
+ * these datums shouldnt have significant behavior, they should just hold data. the lists are filled and emptied by the subsystem.
+ */
+/datum/spatial_grid_cell
+	///our x index in the list of cells. this is our index inside of our row list
+	var/cell_x
+	///our y index in the list of cells. this is the index of our row list inside of our z level grid
+	var/cell_y
+	///which z level we belong to, corresponding to the index of our gridmap in SSspatial_grid.grids_by_z_level
+	var/cell_z
+	//every data point in a grid cell is separated by usecase
+
+	//when empty, the contents lists of these grid cell datums are just references to a dummy list from SSspatial_grid
+	//this is meant to allow a great compromise between memory usage and speed.
+	//now orthogonal_range_search() doesnt need to check if the list is null and each empty list is taking 12 bytes instead of 24
+	//the only downside is that it needs to be switched over to a new list when it goes from 0 contents to > 0 contents and switched back on the opposite case
+
+	///every hearing sensitive movable inside this cell
+	var/list/hearing_contents
+	///every client possessed mob inside this cell
+	var/list/client_contents
+
+/datum/spatial_grid_cell/New(cell_x, cell_y, cell_z)
+	. = ..()
+	src.cell_x = cell_x
+	src.cell_y = cell_y
+	src.cell_z = cell_z
+	//cache for sanic speed (lists are references anyways)
+	var/list/dummy_list = SSspatial_grid.dummy_list
+
+	if(length(dummy_list))
+		dummy_list.Cut()
+		stack_trace("SSspatial_grid.dummy_list had something inserted into it at some point! this is a problem as it is supposed to stay empty")
+
+	hearing_contents = dummy_list
+	client_contents = dummy_list
+
+/datum/spatial_grid_cell/Destroy(force, ...)
+	if(force)//the response to someone trying to qdel this is a right proper fuck you
```

Comment:
stack trace, I wish to be able to yell at people if they do it in ci

Reply:
done